### PR TITLE
Indicator functionality

### PR DESCRIPTION
**Pull request by artemboyko43**

### CODE REVIEW

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php [hunk 1]
```diff
@@ -152,4 +155,113 @@ public function transformUrn(string $dataset_urn): string {
...
+
+    if ($full) {
+      $dataset_structure = $this->getDasetStructure($urn, TRUE);
+      if ($dataset_structure['data']['glossaries']) {
+        $dataset_glossaries = $dataset_structure['data']['glossaries'];
+
+        $indicators_items = [];
+        foreach ($dataset_glossaries as $dataset_glossary) {
+          // @todo check ability to implement better condition.
+          if (str_contains($dataset_glossary['id'], 'INDICATOR')) {
```
Comment (illya-havsiyevych) [inline modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php]: hardcode ?

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php [hunk 2]
```diff
@@ -152,4 +155,113 @@ public function transformUrn(string $dataset_urn): string {
...
+   *   The dataset indicators list.
+   */
+  public function datasetIndicators(string $urn, bool $full = FALSE, array $selected_indicators = []) {
+    $avaiability_data = $this->datasetAvaiability($urn);
+    $indicators = [];
+
+    if (empty($selected_indicators) && !empty($avaiability_data['data']['dataConstraints'][0]['cubeRegions'][0]['memberSelection'])) {
+      $member_selection = $avaiability_data['data']['dataConstraints'][0]['cubeRegions'][0]['memberSelection'];
+      foreach ($member_selection as $value) {
+        if ($value['componentId'] == 'INDICATOR') {
```
Comment (illya-havsiyevych) [inline modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php]: hardcode ?

File: modules/quanthub_indicator/src/QuanthubIndicatorContentEntityTrackingManager.php [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+   * only be recognized if an appropriate @code $entity->original @endcode value
+   * is set.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   The content entity to be indexed.
+   * @param bool $new
+   *   (optional) TRUE if this is a new entity, FALSE if it already existed (and
+   *   should already be known to the tracker).
+   */
+  public function trackEntityChange(ContentEntityInterface $entity, bool $new = FALSE) {
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/QuanthubIndicatorContentEntityTrackingManager.php:61]: let's refactor this one to reduce total size, for example:
- introduce helper methods over *_translations, *_ids
- helper methods over each block in foreach over if ($*_ids) {

File: modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Process the field.
+   */
+  protected function processField(FieldInterface $field) {
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php:169]: let's refactor this one to reduce total size
- let's introduce helper methods for each case in switch like 'rendered_item':

File: modules/quanthub_indicator/src/Plugin/search_api/datasource/QuanthubContentEntity.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ *   id = "entity",
+ *   deriver = "Drupal\search_api\Plugin\search_api\datasource\ContentEntityDeriver"
+ * )
+ */
+class QuanthubContentEntity extends ContentEntity {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function loadMultiple(array $ids) {
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Plugin/search_api/datasource/QuanthubContentEntity.php:20]: it looks like this method is doing 2 tasks - 2 big foreach
consider refactoring by splitting

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+    if ($processed_ids) {
+      if ($this->hasValidTracker()) {
+        $this->getTrackerInstance()->trackItemsIndexed($processed_ids);
+      }
+      // Since we've indexed items now, triggering reindexing would have some
+      // effect again. Therefore, we reset the flag.
+      $this->setHasReindexed(FALSE);
+
+      $description = 'This hook is deprecated in search_api:8.x-1.14 and is removed from search_api:2.0.0. Please use the "search_api.items_indexed" event instead. See https://www.drupal.org/node/3059866';
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Entity/QuanthubIndex.php:157]: I'm worried about this description

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    // Return the loaded items.
+    return $items;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function indexSpecificItems(array $search_objects) {
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Entity/QuanthubIndex.php:83]: too big, consider refactoring

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    }
+
+    // Remember the items that were initially passed, to be able to determine
+    // the items rejected by alter hooks and processors afterwards.
+    $rejected_ids = array_keys($items);
+    $rejected_ids = array_combine($rejected_ids, $rejected_ids);
+
+    // Preprocess the indexed items.
+    $this->alterIndexedItems($items);
+    $description = 'This hook is deprecated in search_api:8.x-1.14 and is removed from search_api:2.0.0. Please use the "search_api.indexing_items" event instead. See https://www.drupal.org/node/3059866';
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Entity/QuanthubIndex.php:115]: I'm worried about this description

File: modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+    $this->languageManager = $languageManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function preprocessIndexItems(array $items) {
+    $this->loadIndicators($items);
+
+    // Annoyingly, this doc comment is needed for PHPStorm. See
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php:141]: WTF ?

File: modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+   *   The string of dataset dimensions.
+   */
+  public function getDimensionsFilterUrl(string $urn) {
+    $dimensions_url = '';
+    if (!empty($this->datasetsDimensions[$urn])) {
+      $dimensions_parts = [];
+      foreach ($this->datasetsDimensions[$urn] as $key => $dimension) {
+        // @todo make configurable INDICATOR.
+        $dimensions_parts[$key] = match ($dimension) {
+          'INDICATOR' => $this->indicatorId,
```
Comment (illya-havsiyevych) [inline modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php]: hardcode ?

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php [hunk 11]
```diff
@@ -38,7 +39,7 @@ class QuanthubSdmxClient {
    * @var array
    */
   private array $headers = [
-    'Accept' => 'application/xml',
+    'Accept' => 'application/json',
```
Comment (illya-havsiyevych) [inline modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php:47]: does this affect all previous functionality ?

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php (hunk 7 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Entity/QuanthubIndex.php:83]: This code comes from parent class from contrib search api module, and don't see the reason and way of refactoring this. All overriding is regarding using _indicator_ divider. So yes this just a small peace of added functionality

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php (hunk 8 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Entity/QuanthubIndex.php:115]: The same answer as upper

File: modules/quanthub_indicator/src/Entity/QuanthubIndex.php (hunk 6 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Entity/QuanthubIndex.php:157]: The same answer as upper

File: modules/quanthub_indicator/src/Plugin/search_api/datasource/QuanthubContentEntity.php (hunk 5 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Plugin/search_api/datasource/QuanthubContentEntity.php:20]: Actually we will not have a big amount of ids here, default ui scenario of reindexing is 50 items, and if we are using drush commands. Drush runs without timeouts and as I remember he works with 50 items too

File: modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php (hunk 4 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php:169]: Added method for rendered_item, don't see the reason for creating methods other fields

File: modules/quanthub_indicator/src/QuanthubIndicatorContentEntityTrackingManager.php (hunk 3 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/QuanthubIndicatorContentEntityTrackingManager.php:61]: Again this method mostly the same as in parent class and I don't want to rewrite it full, for easy updating in future is search api will be updated

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php (hunk 1 above)
Comment (artemboyko43) [reply modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php]: Solved

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php (hunk 2 above)
Comment (artemboyko43) [reply modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php]: solved

File: modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php (hunk 10 above)
Comment (artemboyko43) [reply modules/quanthub_indicator/src/Plugin/search_api/processor/QuanthubIndicatorProcessor.php]: solved

File: modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php (hunk 11 above)
Comment (artemboyko43) [reply modules/quanthub_sdmx_sync/src/QuanthubSdmxClient.php:47]: No this will not affect anything, in sdmxClient I use only json_decode

**Review by illya-havsiyevych — APPROVED** [review]

**Pull request merged**